Run systemtests under parent python thread by default

clayton-tom opened the pull request:
**Description of work.**
Previously all systemtests were run under an individual subprocess, making it slightly more involved to debug them.
Now, by default, the selected tests are run under the parent python thread, unless the user specifies the `-j n` option which would run the tests as before using `n` cores.

This PR also adds a section to the systemtesting dev docs with some instructions on setting up a Pycharm configuration to debug a set of system tests there.

**To test:**
1. From the command line, run a few Systemtests using `-j n` option to make sure the previous functionality still works as it did
2. Run them again without `-j` , make sure you're getting the expected output and that things run smoothly
3. Pick your favourite system test, fiddle with it or the code it runs such that it fails in a couple of different ways, and make sure that this a) does not break the running of all of the tests and b) gives you some sort of sensible output
4. Try debugging this test in Pycharm using the instructions, making sure the instructions are correct and easy enough to follow.

Fixes #31149 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

*This does not require release notes* because **it is a change only devs will see**

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Run SystemTests under parent Python thread by default RE #31149

Commit:
Add section to systemtest dev-docs with debug instructions RE #31149

martyngigg commented on Testing/SystemTests/lib/systemtests/systemtesting.py:
```diff
@@ -667,6 +671,46 @@ def start(self, script):
         os.remove(tmp_file.name)
         return results
 
+    def start_in_parent_thread(self, script):
```
Can we rename this `start_in_current_process`? I think it's a more accurate description of what's happening?

clayton-tom replied:
Good point, will do

martyngigg commented on Testing/SystemTests/lib/systemtests/systemtesting.py:
```diff
@@ -667,6 +671,46 @@ def start(self, script):
+        exec_globals = dict()
+        exec_locals = dict()
+        exitcode = None
+        dual_stdout = DualStdOut()
```
Can I just make sure what this is used for? We want to run the test and both print the output to the screen but also capture it to return to be store later?

clayton-tom replied:
Yes that's right, it's an attempt the replicate the functionality from the Process object that gives you a dump of what that process (running a test) printed to stdout in order to collate the results

martyngigg commented on Testing/SystemTests/lib/systemtests/systemtesting.py:
```diff
@@ -667,6 +671,46 @@ def start(self, script):
+            detail = ex.args[0]
+            cl, exc, tb = sys.exc_info()
+            line_number = traceback.extract_tb(tb)[-1][1]
+        print("%s at line %d of SystemTest for %s.%s: %s" % (error_class, line_number, script._modname,
```
Can we use an `f-string` here?

martyngigg commented on Testing/SystemTests/lib/systemtests/systemtesting.py:
```diff
@@ -700,6 +744,12 @@ def asString(self, clean=False):
         return code
 
+    def asStringSingleThread(self, clean=False):
+        code = self.asString(clean)
```
Would it be worth just adding a parameter to `asString` to not add the `sys.exit` rather than having to trim it off later?

martyngigg reviewed (changes requested):
This is going to be a very useful addition for developers being able to debug the system tests. Nice work. Just some minor comments to think about.

As comment against one of the functions, can we replace everwhere it says parent thread/current Python thread with current Python process? It describes the situation a bit more clearly as for the non-multiprocess case we are not spawning anything else threads or processes.

Commit:
Changes from review comments RE #30048

martyngigg commented on Testing/SystemTests/lib/systemtests/systemtesting.py:
```diff
@@ -1023,7 +1072,7 @@ def __shouldTest(self, suite):
         return True
 
-    def executeTests(self, tests_done=None):
+    def executeTests(self, tests_done=None, status_dict=None):
```
In/out parameters tend to be avoided in Python as, unlike C++ (where they still should be minimized):
  - it's hard to tell whether the intention of the author was that it should be modified or that this was an accident 
  - returning multiple values from a Python function is trivial.

I know that `tests_done` is done like this too but can we avoid adding further arguments used like this? Can a local `status_dict` be created in this function and returned to the caller. The caller can then update their main dictionary with these results?

martyngigg commented on dev-docs/source/SystemTests.rst:
````diff
@@ -236,6 +236,32 @@ properties, go to ```Command Arguments``` and append flags as appropriate.
+This is found in ``/Testing/SystemTests/Scripts/runSystemTests.py``
+
+The parameters for the configuration can be set just like the command line
+args when running the tests from the ``systemtest.bat`` script, e.g pass
````
On macOS/Linux the script is unfortunately called `systemtest` so could we say `systemtest.bat/systemtest` in script to be clear this is not Windows only?

martyngigg reviewed (changes requested):
This works brilliantly from Pycharm with the instructions provided and I can set a breakpoint in a system test finally!

Just a couple of minor code comments and then this is ready to go.

Commit:
Changes from comments RE #31149

Remove in/out parameter in favour of return, and proper name systemtest script in docs

martyngigg reviewed (approved):
Thanks for making these changes. This is good to go!
